### PR TITLE
Cli: `run-testenv`

### LINKED ISSUE #5911: CLI: `run-testenv`
For testing

### PR DESCRIPTION
**Pull request by TimeEngineer**
depends on #5898 
closes #5911

### CODE REVIEW

File: cli/run_testenv.fish [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+# Cross-shell script directory detection
+# set SCRIPT_DIR (dirname (realpath -s $argv[1]))
+set SCRIPT_DIR (dirname (realpath (status -f)))
+
+# In Python we trust (aka shell's tempfile&mktemp doesn't work on all platforms)
+set SOURCE_FILE (python -c "import tempfile; print(tempfile.mkstemp()[1])")
+
+# Run python script and source
+eval cargo run --package parsec_cli --features testenv run-testenv --main-process-id $$ --source-file "$SOURCE_FILE" $argv; or exit $status
```
Comment (FirelightFlagboy) [inline cli/run_testenv.fish:27]: The comment seems misleading ;)

File: cli/run_testenv.sh [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    else
+        export PRE_COMMIT_HOME="$XDG_CACHE_HOME/pre-commit"
+    fi
+fi
+
+# In Python we trust (aka shell's tempfile&mktemp doesn't work on all platforms)
+SOURCE_FILE=$(python -c "import tempfile; print(tempfile.mkstemp()[1])")
+
+# Run python script and source
+cargo run --package parsec_cli --features testenv run-testenv --main-process-id $$ --source-file "$SOURCE_FILE" $@ || return $?
```
Comment (FirelightFlagboy) [inline cli/run_testenv.sh:33]: The comment is misleading

File: cli/src/main.rs [hunk 3]
```diff
@@ -23,10 +25,16 @@ enum Command {
     CreateOrganization(create_organization::CreateOrganization),
     /// List all devices
     ListDevices(list_devices::ListDevices),
+    #[cfg(feature = "testenv")]
+    /// Create a temporary environment and initialize a test setup for parsec.
+    /// #### WARNING: it also leaves an in-memory backend running in the background on port `6770`.
```
Comment (FirelightFlagboy) [inline cli/src/main.rs:30]: I didn't see a cleanup for the in-memory server, that's expected ?

Comment (FirelightFlagboy) [inline cli/src/main.rs:30]:
```suggestion
    /// #### WARNING: it also leaves an in-memory server running in the background on port `6770`.
```

File: cli/src/run_testenv.rs [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+    let organization_id: OrganizationID = "Org".parse().expect("Unreachable");
+
+    // Create organization
+    let bootstrap_token = loop {
+        match create_organization_req(&organization_id, &addr, DEFAULT_ADMINISTRATION_TOKEN).await {
+            Ok(bootstrap_token) => break bootstrap_token,
+            Err(e) => eprintln!("{e}"),
+        }
+        // Need to wait until testbed server starts
+        std::thread::sleep(std::time::Duration::new(1, 0));
```
Comment (FirelightFlagboy) [inline cli/src/run_testenv.rs:205]: I would prefer to have this step in `new_env` when we start the `testbed` server.

Comment (FirelightFlagboy) [inline cli/src/run_testenv.rs:205]: I would prefer a stronger check that wait for a specific output on the stdout of the testbed server.

Comment (TimeEngineer) [reply cli/src/run_testenv.rs:205]: But how can I detect that the testbed isn't start ? :thinking:

File: cli/src/main.rs (hunk 3 above)
Comment (TimeEngineer) [reply cli/src/main.rs:30]: It is closed with your terminal

**Review by FirelightFlagboy — APPROVED** [review]

File: cli/src/run_testenv.rs (hunk 4 above)
Comment (vxgmichel) [reply cli/src/run_testenv.rs]: A classic trick is to wait for the server to print to stdout, that's what we did in `run_testenv.py`:
https://github.com/Scille/parsec-cloud/blob/b4631f0562b6a9f2f5ce9a51e5e49f8df4421dd1/tests/scripts/run_testenv.py#L194-L198

Can't you do the same thing here?

File: server/tests/scripts/run_testbed_server.py [hunk 5]
```diff
@@ -169,7 +169,8 @@ async def test_drop(organization_id: str):  # type: ignore[misc]
             print(
                 f"{BOLD_YELLOW}All set !"
                 f"{NO_COLOR} Don't forget to export `TESTBED_SERVER_URL` environ variable:\n"
-                f"{PINK}export TESTBED_SERVER_URL='parsec://127.0.0.1:{port}?no_ssl=true'{NO_COLOR}"
+                f"{PINK}export TESTBED_SERVER_URL='parsec://127.0.0.1:{port}?no_ssl=true'{NO_COLOR}",
+                flush=True,
```
Comment (FirelightFlagboy) [inline server/tests/scripts/run_testbed_server.py:173]: 🫣 That's why we don't have output last time :sweat:

**Review by FirelightFlagboy — APPROVED** [review]

File: server/tests/scripts/run_testbed_server.py (hunk 5 above)
Comment (TimeEngineer) [reply server/tests/scripts/run_testbed_server.py:173]: Yes, @vxgmichel found it

File: cli/src/run_testenv.rs (hunk 4 above)
Comment (TimeEngineer) [reply cli/src/run_testenv.rs]: Done

**Review by FirelightFlagboy — APPROVED** [review]

**Review by vxgmichel — APPROVED** [review]
It would be nice to add a couple of CLI options and help to the command. As a reference, the old scripts had the following help message:
```
± source tests/scripts/run_testenv.sh --help 
Usage: run_testenv.py [OPTIONS]

  Create a temporary environment and initialize a test setup for parsec.

  WARNING: it also leaves an in-memory backend running in the background on
  port 6888.

  It is typically a good idea to source this script in order to export the XDG
  variables so the upcoming parsec commands point to the test environment:

          $ source tests/scripts/run_testenv.sh

  This scripts create two users, Alice and Bob who both own two devices,
  laptop and pc. They each have their workspace, respectively alice_workspace
  and bob_workspace, that their sharing with each other.

  The --empty (or -e) argument may be used to bypass the initialization of the
  test environment:

          $ source tests/scripts/run_testenv.sh --empty

  This can be used to perform a user or device enrollment on the same machine.
  For instance, consider the following scenario:

          $ source tests/scripts/run_testenv.sh
          $ parsec core gui
          # Connect as bob@laptop and register a new device called pc
          # Copy the URL

  Then, in a second terminal:

          $ source tests/scripts/run_testenv.sh --empty
          $ xdg-open "<paste the URL here>"  # Or
          $ firefox --no-remote "<paste the URL here>"
          # A second instance of parsec pops-up
          # Enter the token to complete the registration

Options:
  -B, --backend-address FROM_URL
  --backend-host TEXT             [default: localhost]
  -p, --backend-port INTEGER      [default: 6888]
  --db TEXT                       [default: MOCKED]
  -b, --blockstore TEXT           [default: MOCKED]
  -P, --password TEXT             [default: test]
  -T, --administration-token TEXT
                                  [default: V8VjaXrOz6gUC6ZEHPab0DSsjfq6DmcJ]
  --email-host TEXT               [default: MOCKED]
  --add-random-users INTEGER      [default: 0]
  --add-random-devices INTEGER    [default: 0]
  -e, --empty
  --cleanup
  -l, --log-level [DEBUG|INFO|WARNING|ERROR|CRITICAL]
                                  [default: WARNING]
  -f, --log-format [CONSOLE|JSON]
                                  [default: CONSOLE]
  -o, --log-file TEXT             [default: stderr]
  --help                          Show this message and exit.

```
In particular, the `--empty` option was very useful to test the invitation process, e.g with one shell running the backend, and another one using an empty env.

But maybe this can be added later, up to you.
LGTM otherwise :+1:

**Pull request merged**